cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+ *
+ * @param nulls_at_start Are the nulls at the start or end?
+ * @param num_rows The number of rows to be rolled over.
+ * @param null_count The number of nulls.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped
+ * iterator construction.
+ */
+struct ungrouped_with_nulls {
+  bool nulls_at_start;
+  cudf::size_type num_rows;
+  cudf::size_type null_count;
+
+  static constexpr bool has_nulls{true};
+  /**
+   * @copydoc ungrouped::row_info
+   */
+  [[nodiscard]] __device__ constexpr cuda::std::
+    tuple<size_type, size_type, size_type, size_type, size_type, size_type, size_type>
+    row_info(size_type i) const noexcept
+  {
+    if (nulls_at_start) {
+      return {null_count, 0, num_rows, 0, null_count, null_count, num_rows};
+    } else {
+      return {null_count,
+              num_rows,
+              null_count,
+              num_rows - null_count,
+              num_rows,
+              0,
+              num_rows - null_count};
+    }
+  }
+};
+
+/**
+ * @brief A group descriptor for a grouped rolling window with nulls
+ *
+ * @param nulls_at_start Are the nulls at the start of each group?
+ * @param labels The group labels, mapping from input rows to group.
+ * @param offsets The group offsets providing the endpoints of each group.
+ * @param null_counts The null counts per group.
+ * @param orderby The orderby column, sorted groupwise.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped
+ * iterator construction.
+ */
+struct grouped_with_nulls {
+  bool nulls_at_start;
+  // Taking raw pointers here to avoid stealing three registers for the sizes which are never
+  // needed.
+  cudf::size_type const* labels;
+  cudf::size_type const* offsets;
+  cudf::size_type const* null_counts;
+
+  static constexpr bool has_nulls{true};
+  /**
+   * @copydoc ungrouped::row_info
+   */
+  [[nodiscard]] __device__ constexpr cuda::std::
+    tuple<size_type, size_type, size_type, size_type, size_type, size_type, size_type>
+    row_info(size_type i) const noexcept
+  {
+    auto const label       = labels[i];
+    auto const null_count  = null_counts[label];
+    auto const group_start = offsets[label];
+    auto const group_end   = offsets[label + 1];
+    if (nulls_at_start) {
+      return {null_count,
+              group_start,
+              group_end,
+              group_start,
+              group_start + null_count,
+              group_start + null_count,
+              group_end};
+    } else {
+      return {null_count,
+              group_start,
+              group_end,
+              group_end - null_count,
+              group_end,
+              group_start,
+              group_end - null_count};
+    }
+  }
+};
+
+/*
+ * Spark requires that orderby columns with floating point type have a
+ * total order on floats where all NaNs compare equal to one-another,
+ * and greater than any non-nan value. These structs implement that logic.
```

Comment:
@mythrocks to confirm that I have this behaviour correct.

Reply:
Yes, that sounds accurate.